File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n=s.size();
        int i=0,j=0;
        int maxi=0;
        unordered_map<char,int>mp;
        while(j<n){
            mp[s[j]]++;
            while(j-i+1>mp.size()){
                mp[s[i]]--;
                if(mp[s[i]]==0) mp.erase(s[i]);
                i++;
            }
            maxi=max(maxi,j-i+1);
            j++;
        }
        return maxi;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        mp={}
        i,n=0,len(s)
        maxi=0
        for j in range(n):
            mp[s[j]]=mp.get(s[j],0)+1
            while(mp[s[j]]>1):
                mp[s[i]]=mp.get(s[i],0)-1
                if mp[s[i]]==0:
                    del mp[s[i]]
                i+=1
            maxi=max(maxi,j-i+1)
            j+=1
        return maxi
        